File: core/domain_foundry_core/receipts.py
"""Plain-language receipts for CLI / Telegram (mirrors app/src/lib/receipts.ts)."""

from __future__ import annotations

from typing import Any


def describe_capture_receipt(receipt: Any, *, pack_titles: dict[str, str] | None = None) -> str:
    """One-line human summary of a CaptureReceipt-like object."""
    titles = pack_titles or {}
    status = getattr(receipt, "status", None) or (
        receipt.get("status") if isinstance(receipt, dict) else None
    )
    routed = getattr(receipt, "routed", None)
    if routed is None and isinstance(receipt, dict):
        routed = receipt.get("routed") or []
    real = []
    for span in routed or []:
        domain = getattr(span, "domain", None) or (
            span.get("domain") if isinstance(span, dict) else None
        )
        if domain and domain not in {"_unfiled", "_ledger"}:
            real.append(span)
    if status == "applied" and len(real) == 1:
        span = real[0]
        domain = getattr(span, "domain", None) or span.get("domain")
        otype = getattr(span, "object_type", None) or span.get("object_type") or "entry"
        human = str(otype).replace("_", " ")
        title = titles.get(str(domain), str(domain))
        article = "an" if human[:1].lower() in "aeiou" else "a"
        return f"Saved to {title} as {article} {human}"
    if status == "applied" and real:
        names = []
        for span in real:
            domain = getattr(span, "domain", None) or span.get("domain")
            names.append(titles.get(str(domain), str(domain)))
        uniq = [n for n in dict.fromkeys(names) if n]
        return f"Saved to {' and '.join(uniq)}" if uniq else "Saved to your journal"
    if status == "review":
        return "Saved for your review — I wasn't fully sure where it belongs."
    if status in {"unfiled", "ledger_only"}:
        return "Kept as unfiled — still saved, not guessed into a passion."
    return f"Captured ({status})."


def pack_install_summary(result: dict[str, Any]) -> dict[str, Any]:
    """Hobby-facing pack install/activate summary — no mesh expert stub."""
    return {
        "name": result.get("name"),
        "title": result.get("title") or (result.get("pack") or {}).get("title"),
        "version": result.get("version"),
    }

File: app/src/lib/receipts.ts
// Plain-language receipts shared by the composer, Today, and Inbox. Internal
// status names stay at this boundary instead of leaking into the product UI.

import type { CaptureReceipt, EntryRow, PackCard } from "./types";

export type ReceiptDescription = {
  tone: "ok" | "unsure" | "error";
  headline: string;
  detail?: string;
  repair: { kind: "refile"; entryId: string } | { kind: "review" } | null;
};

function packTitle(packs: PackCard[], domain: string | null | undefined): string {
  if (!domain || domain === "_unfiled" || domain === "_ledger") return "";
  return packs.find((pack) => pack.name === domain)?.title ?? domain;
}

function humanType(objectType: string | null | undefined): string {
  return (objectType ?? "entry").replace(/_/g, " ");
}

function article(word: string): string {
  return /^[aeiou]/i.test(word) ? "an" : "a";
}

export function describeReceipt(receipt: CaptureReceipt, packs: PackCard[]): ReceiptDescription {
  const real = receipt.routed.filter(
    (span) => span.domain && span.domain !== "_unfiled" && span.domain !== "_ledger",
  );
  const degraded = receipt.llm_error
    ? "Model routing was unavailable, so keyword rules did the filing — check Settings → Providers."
    : undefined;

  switch (receipt.status) {
    case "applied": {
      if (real.length === 1) {
        const span = real[0];
        const type = humanType(span.object_type);
        return {
          tone: "ok",
          headline: `Saved to ${packTitle(packs, span.domain)} as ${article(type)} ${type}`,
          detail: degraded,
          repair: null,
        };
      }
      const names = [...new Set(real.map((span) => packTitle(packs, span.domain)))].filter(Boolean);
      return {
        tone: "ok",
        headline: names.length ? `Saved to ${names.join(" and ")}` : "Saved to your journal",
        detail: degraded,
        repair: null,
      };
    }
    case "review":
      return {
        tone: "unsure",
        headline: "Saved — waiting for your OK before it changes anything",
        detail: degraded ?? "You'll find it in Inbox.",
        repair: { kind: "review" },
      };
    case "unfiled":
      return {
        tone: "unsure",
        headline: "Saved — I wasn't sure where this belongs",
        detail: degraded ?? "File it from Inbox in one click.",
        repair: { kind: "refile", entryId: receipt.entry_id },
      };
    case "ledger_only":
    default:
      return {
        tone: "unsure",
        headline: "Saved to your journal",
        detail: degraded ?? "Install or create a passion and entries like this get filed automatically.",
        repair: null,
      };
  }
}

export function describeRow(row: EntryRow, packs: PackCard[]): ReceiptDescription {
  const fake: CaptureReceipt = {
    entry_id: row.id,
    capture_event_id: row.capture_event_id,
    status: row.status,
    routed: [
      {
        domain: row.domain,
        object_type: row.object_type,
        operation: row.operation,
        disposition: row.status,
        confidence: row.routing_confidence,
      },
    ],
    projection_status: "n/a",
    idempotent_replay: false,
    summary: row.summary,
    llm_error: null,
    domain_hint: null,
  };
  return describeReceipt(fake, packs);
}
